storage/leia.go

```diff
@@ -0,0 +1,300 @@
+}
+
+func (k *kvBackedLeiaStore) JSONLDCollection(name string) leia.Collection {
+	config, ok := k.collectionConfigSet[name]
+	if !ok {
+		// we panic here because this is a programming error, not a runtime error
+		panic("JSON collection not configured")
+	}
+	if config.CollectionType != JSONLDCollectionType {
+		// we panic here because this is a programming error, not a runtime error
+		panic("Incorrect collection configuration")
+	}
+	underlying := kvBackedCollection{
+		backup:     k.backup,
+		config:     config,
+		underlying: k.store.JSONLDCollection(name),
+	}
+	return underlying
+}
+
+func (k *kvBackedLeiaStore) Close() error {
+	return k.store.Close()
+}
+
+func (k kvBackedCollection) AddIndex(index ...leia.Index) error {
+	return k.underlying.AddIndex(index...)
+}
+
+func (k kvBackedCollection) DropIndex(name string) error {
+	return k.underlying.DropIndex(name)
+}
+
+func (k kvBackedCollection) NewIndex(name string, parts ...leia.FieldIndexer) leia.Index {
+	return k.underlying.NewIndex(name, parts...)
+}
+
+func (k kvBackedCollection) Add(jsonSet []leia.Document) error {
+	// first in backup
+	for _, doc := range jsonSet {
+		ref := k.Reference(doc)
+
+		if err := k.backup.WriteShelf(context.Background(), k.config.BackupShelf, func(writer stoabs.Writer) error {
+			return writer.Put(stoabs.BytesKey(ref), doc)
+		}); err != nil {
+			return err
+		}
+	}
+	// then in index
+	return k.underlying.Add(jsonSet)
+}
+
+func (k kvBackedCollection) Get(ref leia.Reference) (leia.Document, error) {
+	return k.underlying.Get(ref)
+}
+
+func (k kvBackedCollection) Delete(doc leia.Document) error {
+	// first in backup
+	ref := k.Reference(doc)
+
+	if err := k.backup.WriteShelf(context.Background(), k.config.BackupShelf, func(writer stoabs.Writer) error {
+		return writer.Put(stoabs.BytesKey(ref), doc)
+	}); err != nil {
+		return err
+	}
+	// then in index
+	return k.underlying.Delete(doc)
+}
+
+func (k kvBackedCollection) Find(ctx context.Context, query leia.Query) ([]leia.Document, error) {
+	return k.underlying.Find(ctx, query)
+}
+
+func (k kvBackedCollection) Reference(doc leia.Document) leia.Reference {
+	return k.underlying.Reference(doc)
+}
+
+func (k kvBackedCollection) Iterate(query leia.Query, walker leia.DocumentWalker) error {
+	return k.underlying.Iterate(query, walker)
+}
+
+func (k kvBackedCollection) IndexIterate(query leia.Query, fn leia.ReferenceScanFn) error {
+	return k.underlying.IndexIterate(query, fn)
+}
+
+func (k kvBackedCollection) ValuesAtPath(document leia.Document, queryPath leia.QueryPath) ([]leia.Scalar, error) {
+	return k.underlying.ValuesAtPath(document, queryPath)
+}
+
+func (k kvBackedCollection) DocumentCount() (int, error) {
+	return k.underlying.DocumentCount()
+}
+
+func (k *kvBackedLeiaStore) backupStorePresent(backupShelf string) bool {
+	backupPresent := false
+
+	_ = k.backup.ReadShelf(context.Background(), backupShelf, func(reader stoabs.Reader) error {
+		isEmpty, err := reader.Empty()
```

Comment:
kinda dangerous to have an error default to backupPresent = false?

Reply:
If it's a technical error, it'll try to fill the backup which also gives an error or it will rewrite some of the same values (hashes as keys).